MSVC: compiler wrapper support

johnwparent opened the pull request:
True compiler wrappers on Windows!

We introduce a true Windows compiler wrapper with this PR! 

In addition to wrapping compiler/linker and behaving much like the current Spack compiler wrapper, this compiler wrapper additionally vendors much improved RPATH support for Windows.

Windows does not have RPaths, loader paths, LD_LIBRARY_PATH, etc, nor any really analogous concept to these features available on other platforms. 

Windows uses a predefined set of heuristics to resolve dynamic library dependencies at runtime, which cannot be easily modified outside of adding to the PATH, which we want to avoid for a number of reasons. 

Our current solution is to take advantage of the fact that Windows' first step in the DL loading heuristic is to look directly adjacent to the exe/DLL requesting the dependency to resolve it, so we create symlinks of all DLLs which a given Spack package depends on and place them in the relevant locations. See [Windows Simulated RPath](https://github.com/spack/spack/blob/develop/lib/spack/llnl/util/filesystem.py#L2449) for more context.

The new compiler wrapper instead directly injects the absolute path of a dependency's DLLs into MSVC import libraries, and anything linking against them simply has the absolute path to the dependency baked in, so all attempts from Windows to resolve that DLL point directly to our Spack based dependency, much like how RPaths work (except there's only one path). 

This PR also disables the current symlink based solution.

The new compiler wrapper additionally vendors support for relocating binaries in and out of the buildcache, dynamically rewriting the paths to the relevant DLLs as they are extracted from the buildcache to point to the locations of their dependencies on the new host machine.

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -84,6 +87,42 @@ def _register(cls):
+        obj_file = env_build_path / "cl.obj"
+        # cleanup object file from build
+        if obj_file.exists():
+            obj_file.unlink()
```
I'm wondering if we might in fact want to make these cl wrappers a small spack package (e.g. their source code lives in a separate repo). That can benefit from things like providing pre-compiled instances of them

johnwparent replied:
That will also save the semi-hacky logic of manually reproducing the build env to compile these and allow us to rely on Spack's installation workflow.

scheibelp commented on lib/spack/spack/compilers/msvc.py:
```diff
@@ -53,14 +53,36 @@ class Msvc(Compiler):
     # Subclasses use possible names of Fortran 90 compiler
     fc_names: List[str] = ["ifx"]
 
+    # There are no rpaths on Windows, so now special arguments should
```
(typo) now -> no

scheibelp commented on lib/spack/spack/compilers/msvc.py:
```diff
@@ -53,14 +53,36 @@ class Msvc(Compiler):
+
+    @property
+    def fc_rpath_arg(self):
+        return ""
```
(side note: this is just used to set the `SPACK_CC_RPATH_ARG` env var, which is used by the wrappers, in some sense we don't need to un-define them, but I do think it's clear)

johnwparent replied:
The arguments set by the base `Compiler` class are incorrect/invalid for MSVC, so I'd rather we unset them then set them to something incorrect, even if they aren't used at the moment to ensure valid modeling. Also I'd rather handle it now than when I introduce the RPATH hack PR.

scheibelp replied:
Coming back to this I realize it would be useful to make the comment more specific:

> The msvc compiler has no analog to `-rpath` in `gcc`

Also, regarding the overall strategy:

Right now it looks like https://github.com/spack/msvc-wrapper/blob/main/cl.cxx capably transmits calls to the underlying compiler; however, right now I don't see it modifying the resulting artifact to help it locate needed libraries. Will it be responsible for that? Or will there be a post-install step for rewriting the binaries.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+#include "cl.hxx"
+
+// String helper methods adding cxx20 features to cxx14
```
This could use a header comment that explains the overall purpose of the file.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+"SPACK_SYSTEM_DIRS"};
+    for(auto &var: SpackEnv)
+        if(!getenv(var.c_str())){
+            throw SpackCompilerContextException(var);
```
I think at the moment the exception message would just be `"SPACK_LINKER_ARG"` if e.g. that wasn't set. This should probably be `{x} isn't set in the environment`

johnwparent replied:
Amended.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+void ToolChainInvocation::cleanupHandles(){
+    CloseHandle(this->procInfo.hProcess);
+    CloseHandle(this->procInfo.hThread);
+}
```
(minor) IIRC if you do resource cleanup in the destructor, then that tends to avoid issues with missing cleanup (in particular since you manage the `ToolchainInvocation` instance with a `unique_ptr`).

johnwparent replied:
Will implement RAII, should have done that to begin with but my C++ was rusty when I wrote this, too used to a GC (I blame Go and Python lol). Revisiting the RPath hack C++ impl with this in mind as well.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+        else{
+            // If it's not c/c++ or fortran we're linking
+            Tool = std::unique_ptr<LdInvocation>(new LdInvocation(command, cli));
+        }
```
Ideally we could explicitly check for `ld`, and other strings would generate an error (which avoids cases where something parses wrong, we get garbage input, and then move directly to `ld`ing).

Also, this is the only place `SupportedTools` is used, I was going to suggest using an `enum`, but I think a function could just convert the command string to an `*Invocation` (or that could be done here without calling out).

johnwparent replied:
We already handle this case. If `command` isn't a string we're expecting we raise a `SpackUnknownCompilerException`. In this conditional, we could _only_ every possible get one of `c/c++`, `intel fortran`, or `link`, any other string will trigger the exception. 

> function could just convert the command string to an *Invocation

I'm a bit confused by this. We're already constructing the `Invocation` classes here, it seems like calling out would just abstract the conditional to elsewhere. I can refactor if that's what you'd like/that would be clearer code, but I just want to be sure we're on the same page here. The "single use" of a method, enum, or this map would apply regardless of the mechanism we're using to determine which tool to invoke so I also don't really see the advantage. Perhaps I'm overlooking something? I personally prefer the map approach here as that's avoiding direct string comp and uses the hash lookup instead, where a function (which presumably would just be `if command == "cl"` etc) or enum (which seems like an uneeded extra step as we'd have to perform string comparison as with the function and then assign the result to an enum reference)  would be slightly more involved.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+    throw SpackCompilerException(cli);
+}
+
+SpackList split(std::string s, std::string delim) {
```
Can you add a "docstring", e.g. explaining what happens if delim is " " and you have `s = "example one"` or `s = "example              two"` (with multiple spaces - in that case do you still get a list of length 2?).

johnwparent replied:
With multiple delims you'd get a list longer than 2. On a second thought, I don't really like that option, and I'd prefer the Python style approach. Will refactor.

scheibelp commented on lib/spack/spack/compilers/msvc.py:
```diff
@@ -264,6 +286,22 @@ def vs_root(self):
+            'cmd /u /c "{}" {} {} {} && set'.format(
+                self.setvarsfile, arch, sdk_ver, "-vcvars_ver=%s" % self.msvc_version
+            ),
+            stderr=subprocess.STDOUT,
```
Is there a reason `stderr` is redirected to `stdout` here?

johnwparent replied:
It's what we're doing in develop right now, I just copied it. I think this is moot since I'll be taking this out once we make the compiler wrapper an independent package.

scheibelp commented on share/spack/win-compiler-wrapper/cl.hxx:
```diff
@@ -0,0 +1,134 @@
+    std::string spackCommand;
+};
+
+class ClInvocation : public ToolChainInvocation {
```
Docstrings might be useful (for example to explain that this represents an invocation of `cl.exe`, vs. "command line invocation").

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+    this->command = command;
+    this->parse_command_args(cli);
+}
+void ToolChainInvocation::interpolateSpackEnv(SpackEnvState &spackenv)
```
For the most part you seem to follow a consistent style, although I'll note that generally you appear to have `SpackEnvState& spackenv` (attaching the reference quality to the type).

johnwparent replied:
It's actually about 50/50 mixed and I personally prefer the reference denotation to the argument.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+        return SpackList();
+}
+
+void loadSpackEnvState(SpackEnvState &spackenv) {
```
IMO this should be a method on `SpackEnvState`.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+}
+
+std::string getSpackEnv(const char * env) {
+    auto envVal = getenv(env);
```
Can you avoid using `auto` in this case? I was trying to reason about whether all possible types for `envVal` would make sense when evaluated in the ternary, but I assume that `getenv` returns `const char*` in all cases (so I'd be curious what the handling is for `""` or `null`.

Since the return type is `std::string`, I'm also unclear how this can return `envVal`

johnwparent replied:
Sure re: `auto`. The handling is to return an empty string. 
Returning a char buffer as a string is legal c++. The string constructor creates a deep copy of the char buffer.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+    size_t matchLen = match.size();
+    if (matchLen > arg.size())
+        return false;
+    return arg.compare(arg.size() - matchLen, matchLen, 0) == 0;
```
I don't see how this works without referring to `match` (just `matchLen`).

johnwparent replied:
Refactor error.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+    return arg.compare(arg.size() - matchLen, matchLen, 0) == 0;
+}
+
+bool endswith(std::string &arg, char const* match){
```
(no change needed) had to look this up to remember that in fact there is no standard library function (or `std::string` function) for this.

johnwparent replied:
There actually is, but it's C++20, which ideally we'd support except I assume we want to support older MSVCs. MSVC 15 (which at the moment is the oldest we support) is up to C++17, which is I believe what I targeted with this.

scheibelp commented on share/spack/win-compiler-wrapper/cl.hxx:
```diff
@@ -0,0 +1,134 @@
+
+#define BUFSIZE 4096
+
+typedef std::vector<std::string> SpackList;
```
IMO `StrList` is a better abbreviation (more clearly indicates this is just for strings).

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,406 @@
+
+std::string ToolChainInvocation::composeLibPathArg(std::string &libPath){
+    return "/LIBPATH:" + libPath;
+}
```
This appears to read in e.g. `SPACK_LINK_DIRS`, but does not add them to the invocation the way the POSIX wrapper does (i.e. turning them into `-L`. Likewise for `SPACK_INCLUDE_DIRS`

johnwparent replied:
See `interpolateSpackEnv` where I call `addExtraLibPaths` on the value read from the env in `loadSpackEnvState` where I then add them to `this->libDirArgs` and then pass that to the command line via `composeCLI`

scheibelp reviewed (changes requested):
Some preliminary requests.

Does this set up PATH or otherwise arrange for Windows builds to use this `cl.exe`?

scheibelp commented on share/spack/win-compiler-wrapper/cl.hxx:
```diff
@@ -0,0 +1,170 @@
+public:
+    using ToolChainInvocation::ToolChainInvocation;
+private:
+    void loadToolchainDependentSpackVars(SpackEnvState& spackenv);
```
(minor) Continuing the theme of https://github.com/spack/spack/pull/38495#discussion_r1367489664, it looks like this is now consistently `type &name` except for here and a few lines above.

scheibelp commented on var/spack/repos/compiler.wrappers/packages/msvc/package.py:
```diff
@@ -0,0 +1,14 @@
+
+
+
+class MsvcCompilerWrapper(Package):
```
Two items:

* Packages that inherit `Package` must implement `install` (I assume we'd need to compile things: perhaps you were waiting on extracting the code into a separate repo, but in the meantime a skeleton install method would be suitable).
* The directory containing the `package.py` file should be derived from the class name like `msvc-compiler-wrapper` (i.e. the dir name is the class name with dashes separating the words)

scheibelp commented on var/spack/repos/compiler.wrappers/repo.yaml:
```diff
@@ -0,0 +1,2 @@
+repo:
+  namespace: compiler.wrapper
```
Interesting idea to maintain this as a separate repo, so e.g. it doesn't show up when listing packages that spack can install (since this is only of use to spack).

I think you'd have to update the config in the bootstrap logic to use it though, and I don't see that part.

johnwparent replied:
Wasn't sure if we'd want to take this approach given our various refactors. Since it seems like this is the way to go, I'll add that.

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -84,6 +87,42 @@ def _register(cls):
+    wrapper_source = wrapper_source_dir / "cl.cxx"
+    env_build_path = pathlib.Path(spack.paths.build_env_path) / "msvc"
+    with working_dir(str(env_build_path)):
+        cl(str(wrapper_source), "/EHsc", extra_env=int_env)
```
I think once this moves to the package.py, the remaining work will be to update `bootstrap.common.ensure_core_dependencies` to make use of it (I think by providing the abstract spec for the wrapper there or in a helper function).

johnwparent replied:
Exactly, that's what I'm doing locally now.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,440 @@
+                                                        {"cl", "c/c++"},
+                                                        {"link", "link"},
+                                                        {"ifort", "intel fortran"},
+                                                        {"ifx", "intel fortran"}
```
IMO using an enum would be cleaner for values.

Also, should we be mapping `ifort` to `fortran` vs. `intel-fortran`? I assume we'll want to support other fortran compilers in the future (that's not something that needs to change here, although I'm curious what the consequences are of making it generic like that here).

johnwparent replied:
I'll switch to the enum. 

> should we be mapping ifort to `fortran` vs. `intel-fortran`

I am of the opinion that since `ifort` is unique to the intel fortran environment, which like MSVC, generally does its own thing, there is value in making the distinction between `intel fortran` and vanilla `fortran` particularly if some specific flags or handling becomes necessary down the line and we need to be able to distinguish between a ported `gfortran` and `ifx`.

scheibelp commented on share/spack/win-compiler-wrapper/cl.cxx:
```diff
@@ -0,0 +1,440 @@
+/**
+ * @file cl.cxx
+ * @author John Parent (john.parent@kitware.com)
+ * @brief A C++ wrapper header file for the MSVC c and c++ compilers and linkers
```
This part is copied from `cl.hxx` and could use minor revising (to be clear I see that not the entire comment is copied, but this paragraph is).

scheibelp reviewed (commented):
I have no major issues with the C++ implementation of things. I make a couple comments on how to construct a Spack package for the wrapper.

haampie commented:
Can you make it part of builtin instead of a separate repo? 

I think it is cleaner to make the wrapper a build dep of (now at least) all packages when platform=windows.

We're probably going that direction with "compilers as deps"

johnwparent commented:
> I think it is cleaner to make the wrapper a build dep of (now at least) all packages when platform=windows.
> 

Is there an efficient way to do this beyond adding `with when("platform=windows"):` to `PackageBase` or the Windows supported build systems?

scheibelp commented:
@alalazo @haampie similar to https://github.com/spack/spack/pull/38495#issuecomment-1815076233, I'm not sure how https://github.com/spack/spack/pull/38495#issuecomment-1807775859 can be addressed without modifying `PackageBase` (or `CMakePackage`+`NMakePackage` + ...); I think that's a reasonable question because I'm also not clear how that approach smooths the transition to compilers-as-deps (both the implementation here and the suggestion would seem to conflict with that).

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -490,6 +493,31 @@ def ensure_gpg_in_path_or_raise() -> None:
+            executables=["cl.exe"], abstract_spec=msvc_compiler_wrapper_root_spec()
+        )
+        # Spec is built, now copy it to the compiler wrapper directory
+        copy(spack.spec.Spec(msvc_compiler_wrapper_root_spec().prefix.bin + "\\cl.exe"), str(msvc_comp_wrapper))
```
(question) why copy the generated wrapper vs. using it in its initial install location? I'm not against maintaining a copy, although:

If the `msvc-compiler-wrapper` package gets new versions (e.g. we find a bug after merging this PR, and want to push a fix that is non-destructive to prior installs) we might have trouble distinguishing whether this copy destination stores the old or new version

scheibelp commented on var/spack/repos/compiler.wrappers/packages/msvc-compiler-wrapper/package.py:
```diff
@@ -0,0 +1,27 @@
+    build_targets = ["cl.exe"]
+
+    def nmake_install_args(self, spec, prefix):
+            return [f"PREFIX={spack.paths.build_env_path}\\msvc"]
```
(typo)

scheibelp reviewed (changes requested):
Overall this looks good in terms of how it retrieves/installs the compiler wrapper as needed. I have a question about how the wrapper itself functions.

Also, IMO it would be good to add a unit-test/ci-check to ensure that the wrapper functions correctly

johnwparent committed:
Spack side MSVC compiler wrapper support

Adds bootstrapping logic for the wrapper

Determine mapping between DLL and lib, provide both to compiler wrapper

buildcache: do not process compiler wrapper components on win

filesystem: ensure dumpbin output parsing is correct

Ensure DLL parsing skips the correct number of intro lines
Use a more specific dll symbol name regex
import libs and DLLs: use a better parsing stop condition

johnwparent committed:
Spack side MSVC compiler wrapper support

Adds bootstrapping logic for the wrapper

Determine mapping between DLL and lib, provide both to compiler wrapper

buildcache: do not process compiler wrapper components on win

filesystem: ensure dumpbin output parsing is correct

Ensure DLL parsing skips the correct number of intro lines
Use a more specific dll symbol name regex
import libs and DLLs: use a better parsing stop condition

Fixup style

johnwparent committed:
Fixup style

johnwparent committed:
filesystem: Correct dumpbin lib parsing

johnwparent committed:
filesystem: Correct dumpbin lib parsing

johnwparent committed:
relocate win rpath: ensure wrapper is aware of install root

To determine when relocation is unneccesary post install (i.e. for another spack dep) we need to know what defines a "spack installed dep".
We do this by informing the wrapper of the spack install prefix root so it can determine if any of the "rpath"'d dependencies have the same prefix.

johnwparent committed:
relocate win rpath: ensure wrapper is aware of install root

To determine when relocation is unneccesary post install (i.e. for another spack dep) we need to know what defines a "spack installed dep".
We do this by informing the wrapper of the spack install prefix root so it can determine if any of the "rpath"'d dependencies have the same prefix.

johnwparent committed:
relocate_win_package: add new env context

Compiler wrapper now supports the SPACK_CONTEXT_ROOT env var to determine root for SFN conversions

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
relocate_win_package: add new env context

Compiler wrapper now supports the SPACK_CONTEXT_ROOT env var to determine root for SFN conversions

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Allow dlls with no exports and exes with exports

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Allow dlls with no exports and exes with exports

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Update relocation handling

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Update relocation handling

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Use new ID system

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Use new ID system

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Finalize id based relocation

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Finalize id based relocation

Signed-off-by: John Parent <john.parent@kitware.com>

style

Signed-off-by: John Parent <john.parent@kitware.com>

style

Signed-off-by: John Parent <john.parent@kitware.com>

no bare return

Signed-off-by: John Parent <john.parent@kitware.com>

Update typing

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
style

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
style

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
no bare return

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Update typing

Signed-off-by: John Parent <john.parent@kitware.com>

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -842,6 +846,16 @@ def add_file(tar: tarfile.TarFile, info: tarfile.TarInfo, path: str):
                 relocate_binaries.append(os.path.relpath(path, prefix))
+                # Windows compiler wrappers are executables all symlinked
+                # to a central executable, we shouldn't relocate any
+                # of them
```
Question: I'm unclear how this applies here (this function is generically about creating a binary cache of any spec, but this comment appears to be about relating the `compiler-wrappers` package specifically, and I don't see any specific testing for that).

e.g. this comment would make more sense to me next to 

```
    elif "pe" in platform.binary_formats and not spec.name != "compiler-wrapper":
        relocate.relocate_msvc_pe_files(binaries, prefix_to_prefix)
```

below (except I would have expected `and spec.name != "compiler-wrapper"`, with no `not`) or

```
if sys.platform == "win32" and not (self.spec.external or self.name == "compiler-wrapper"):
```

in `package_base`/`windows_establish_runtime_linkage`

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -2107,7 +2123,7 @@ def install_root_node(
         tty.msg('Installing "{0}" from a buildcache'.format(spec.format()))
         extract_tarball(spec, tarball_stage, force)
-        spec.package.windows_establish_runtime_linkage()
+
```
vestigial space

scheibelp commented on lib/spack/spack/bootstrap/_common.py:
```diff
@@ -205,14 +206,15 @@ def _executables_in_store(
+            searchable_paths = [str(x) for x in prefix.glob("**")]
+            binary = spack.util.executable.which_string(*executables, path=searchable_paths)
+            if binary:
+                bin_dir = os.path.dirname(binary)
```
I think this change is to account for the fact that binaries on Windows are found outside the `bin/` directory. It might be worth making that distinction explicit with `if sys.platform == "win32"`

johnwparent replied:
Makes sense to me.

johnwparent replied:
Added a comment about why this may be necessary specifically for Windows. The code still applies to and is valid for all platforms.

scheibelp replied:
I think it would be good to keep the old behavior on non-Windows: if bootstrapped specs are external, searching all directories recursively can be expensive.

scheibelp commented on lib/spack/spack/bootstrap/core.py:
```diff
@@ -495,6 +495,23 @@ def ensure_gpg_in_path_or_raise() -> None:
 
+def msvc_compiler_wrapper_root_spec() -> str:
+    """Return the root spec for the MSVC compiler wrapper package"""
+    return _root_spec("compiler-wrapper@0.1.0")
```
compiler-wrapper is inserted as a dependency now, so is this still required? It doesn't pin a version (but I think we would want to install the latest version).

johnwparent replied:
It is a run/link dep right? If so, you're right, we can pull this out, otherwise, it needs to stay so we can relocate from a binary cache without needing to do a local install first.

scheibelp replied:
No: build type (see `asp.py`/`SpackSolverSetup.define_runtime_constraints`)

Makes sense though about needing to ensure it's available for relocation, but I think that means that we only need `ensure_msvc_relocate_or_raise`, not `ensure_msvc_compiler_wrappers_or_raise`

scheibelp commented on lib/spack/spack/package_base.py:
```diff
@@ -131,11 +132,8 @@ def windows_establish_runtime_linkage(self):
-            win_rpath.add_rpath(*self.win_add_rpath())
-            win_rpath.establish_link()
+        if sys.platform == "win32" and not (self.spec.external or self.name == "compiler-wrapper"):
+            ssys.relocate_win_rpath(self)
```
Question: are other no-relocation scenarios expected? Maybe it should be possible to disable this with a flag at the package definition level (perhaps I'm overly-concerned about the fragility of referring to this by name - we already depend on that in `asp.py` so perhaps not critical to address here).

johnwparent replied:
The only other case where we'd want to avoid relocation is if a library was built purely statically or is purely plugin based where the plugins themselves do not have dependencies. In either scenario, skipping the relocation would be an optimization, as the python code + the tool will skip those scenarios anyway. 

I'm happy to support a flag, but I don't see it being very used.

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +60,210 @@ def fix_darwin_install_name(path: str) -> None:
+
+    get_binary_file_from_prefix = lambda x: glob.glob(
+        os.path.join(package.spec.prefix, f"**\\*.{x}"), recursive=True
+    )
```
(minor) slightly more idiomatic to use `llnl.util.filesystem.find`.

johnwparent replied:
Yes, should have used that.

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +60,210 @@ def fix_darwin_install_name(path: str) -> None:
+    for search_dir in dirs_to_relocate:
+        for entry in os.scandir(search_dir):
+            if not (entry.is_symlink() or entry.is_junction()) and entry.is_dir():
+                dirs_to_relocate.append(entry.path)
```
This "looks right" but IMO would benefit from being extracted into a `get_dirs_recursive` function (also because `relocate_msvc_pe_files` appears to use the same code).

scheibelp commented on lib/spack/spack/relocate.py:
```diff
@@ -35,10 +37,75 @@ def _patchelf() -> Optional[executable.Executable]:
+    import spack.bootstrap
+
+    if sys.platform != "win32":
+        return None
```
(minor) IMO you can raise an exception (I think things that call this function should make sure they are in a context where it makes sense to do so, and I think that is already the case, but let me know if I'm missing something where returning `None` smooths things over).

johnwparent replied:
I tend to a agree, I was probably just being paranoid.

johnwparent committed:
Better attempts to capture error

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Better attempts to capture error

Signed-off-by: John Parent <john.parent@kitware.com>

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +60,210 @@ def fix_darwin_install_name(path: str) -> None:
+    db_bin_dir = os.path.dirname(pkg["msvc"].cc)
+    dumpbin = which("dumpbin", path=db_bin_dir)
+    if not dumpbin:
+        raise RuntimeError("dumpbin required, but not found")
```
(minor) you can set `required=True` on `which`.

johnwparent replied:
Yes, thanks

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +60,210 @@ def fix_darwin_install_name(path: str) -> None:
+    """Extracts the string at ID 59673 from the string table in a given dll
+    Arguments:
+        lib: the dll to extract the string resource from
+    Returns the string at ID 59673 if one is present, None otherwise
```
(question) I was hoping one side effect of using a novel resource type in https://github.com/spack/msvc-wrapper/pull/23 was that this was no longer "magic-number-dependent" (i.e. could just be extracting from ID 1 of the spack resource type). I see that PR hasn't been updated yet though - will this change when that changes?

johnwparent replied:
Comment updated for new behavior.

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +60,210 @@ def fix_darwin_install_name(path: str) -> None:
+
+
+def collect_pe_api(pkg, pe):
+    regex = re.compile(".*? ([?]*[@$?a-zA-Z0-9_]+)(?: = ([a-zA-Z_][a-zA-Z0-9_]*))?\r$")
```
Can you add a comment explaining the pattern this is looking for?

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +60,210 @@ def fix_darwin_install_name(path: str) -> None:
+        tty.debug(f"Cannot verify library {lib} as COFF.")
+    if relocate_exe.returncode == 0:
+        return True
+    return False
```
(minor) `return relocate_exe.returncode == 0`

Also question: is this to filter out things which look like libraries by name but are not? Or is this in fact unexpected? i.e. would it make sense to *not* suppress a `raise` here (or turn it into something more understandable)?

johnwparent replied:
Right, this is to filter things that look like import libraries but may not be (i.e. standard static libraries). 
It could make sense to capture the output of my tool and pipe that to the user.

scheibelp commented on lib/spack/spack/relocate.py:
```diff
@@ -35,10 +37,75 @@ def _patchelf() -> Optional[executable.Executable]:
+    ev.set_path("SPACK_RELOCATE_PATH", dirs_to_relocate)
+    dll_lib_map = {}
+    for target in targets:
+        # we relocate exes and dlls, libs are regenerated from themselves
```
I'm not sure what "from themselves" is adding (it might be important but what I mean is that I missed the importance - maybe it's that libs can be regenerated into something correct with no extra information?).

johnwparent replied:
We don't actually go through the lib file and re-write any dll references, instead we recreate the .lib from the existing copy. I'll rephrase.

scheibelp commented on lib/spack/spack/relocate.py:
```diff
@@ -35,10 +37,75 @@ def _patchelf() -> Optional[executable.Executable]:
+    dll_lib_map = {}
+    for target in targets:
+        # we relocate exes and dlls, libs are regenerated from themselves
+        # if import libs, if static, no op
```
I'm having trouble parsing this line.

johnwparent replied:
I should have said "if .lib, if static..." the intent was to differentiate handling static libraries vs import libraries

scheibelp commented on lib/spack/spack/relocate.py:
```diff
@@ -35,10 +37,75 @@ def _patchelf() -> Optional[executable.Executable]:
+        # via the libs to determine the proper association
+        if target.endswith(".lib"):
+            if sfs.verify_import_lib(target):
+                regex = re.compile("|".join(re.escape(p) for p in prefixes.keys()))
```
question: should there be any prefix normalization (e.g. removal of final separator)?

johnwparent replied:
The final `|`? The wrapper should handle that, but it's better if we ensure it doesn't have to.

johnwparent replied:
Ah I misread where this code was initially. I think this is good as is.

scheibelp replied:
I meant if the path itself ends with a path separator like `\`. I think `get_importlib_target` will always return a file though (never a directory); if that's right then a comment might be helpful.

scheibelp commented on lib/spack/spack/relocate.py:
```diff
@@ -35,10 +37,75 @@ def _patchelf() -> Optional[executable.Executable]:
+                        "in this prefix, skipping relocation..."
+                    )
+    for target in targets:
+        # now we process the DLLs, now that we've mapped all our libs
```
expanding on this comment slightly would be useful: `update DLLs/exes to replace x with y using prior info z"

scheibelp commented on lib/spack/spack/relocate.py:
```diff
@@ -299,6 +366,10 @@ def is_elf_magic(magic: bytes) -> bool:
 
 
+def is_msvc_magic(magic: bytes) -> bool:
+    return magic.startswith(b"!<arch>\n") or magic.startswith(b"MZ")
```
A couple questions:

* The pattern is generally `is_<filetype>_magic`: would something like `is_pe_magic` be accurate?`
* Is there a more-specific way to identify files than `MZ` (two bytes seems like it might misidentify files we aren't interested in)

johnwparent replied:
This covers both PE and coff, but I'll come up with a different term than "msvc".

> Is there a more-specific way to identify files than MZ (two bytes seems like it might misidentify files we aren't interested in)

I'll look into it, but those are the "magic bytes" according to Windows. We can always do what we do for the coff files and implement some deeper C++ analysis.

johnwparent replied:
There's another signature we can validate at bytes 60-64. Added that functionality.

johnwparent committed:
wip

Signed-off-by: John Parent <john.parent@kitware.com>

cleanup comments

Signed-off-by: John Parent <john.parent@kitware.com>

style

Signed-off-by: John Parent <john.parent@kitware.com>

Better guard against non compiler packages

Signed-off-by: John Parent <john.parent@kitware.com>

Review

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
wip

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
More graceful read of magic bytes

Signed-off-by: John Parent <john.parent@kitware.com>

style

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Relocation type gate

Signed-off-by: John Parent <john.parent@kitware.com>

revert me: try to force fresh

Signed-off-by: John Parent <john.parent@kitware.com>

proper module

Signed-off-by: John Parent <john.parent@kitware.com>

revert em

Signed-off-by: John Parent <john.parent@kitware.com>

bump core bootstrap compiler wrapper version

Signed-off-by: John Parent <john.parent@kitware.com>

is_junction is too new

3.12 and on unfortunately

Signed-off-by: John Parent <john.parent@kitware.com>

scheibelp commented on lib/spack/spack/package_base.py:
```diff
@@ -131,11 +132,17 @@ def windows_establish_runtime_linkage(self):
-            win_rpath.add_rpath(*self.win_add_rpath())
-            win_rpath.establish_link()
+        if sys.platform == "win32" and (
+            self.spec.satisfies("%compiler-wrapper") and self.name != "compiler-wrapper"
```
`compiler-wrapper` doesn't depend on itself: this check can be truncated

scheibelp commented on lib/spack/spack/relocate.py:
```diff
@@ -35,10 +37,77 @@ def _patchelf() -> Optional[executable.Executable]:
 
 
+def relocate_msvc_pe_files(targets, prefixes: dict):
+
```
extra space here, but more-importantly: docstring

scheibelp commented on lib/spack/spack/relocate.py:
```diff
@@ -35,10 +37,77 @@ def _patchelf() -> Optional[executable.Executable]:
+    for search_dir in dirs_to_relocate:
+        for entry in os.scandir(search_dir):
+            if not (entry.is_symlink() or entry.is_junction()) and entry.is_dir():
+                dirs_to_relocate.append(entry.path)
```
This appears to be collecting (exactly) one additional layer of directories: is there some reason this isn't transitive (or conversely, why does it need the extra layer - I'd assume passing in the base prefix would be enough to figure out what to relocate)?

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +61,176 @@ def fix_darwin_install_name(path: str) -> None:
+    try:
+        relocate_exe("--coff", lib, "--verify", ignore_errors=[1])
+    except ProcessError:
+        tty.debug(f"Cannot verify library {lib} as COFF.")
```
I agree with https://github.com/spack/spack/pull/38495#discussion_r3184436231 that it would be useful to collect stdout/stderr and print them in this case.

scheibelp commented on lib/spack/spack/relocate.py:
```diff
@@ -35,10 +37,77 @@ def _patchelf() -> Optional[executable.Executable]:
+                else:
+                    tty.debug(
+                        f"Import lib: {target} does not reference a DLL"
+                        "in this prefix, skipping relocation..."
```
I think "in this prefix" means "does not match any of the relocation prefixes"

it might be worth inserting those into the debug message.

I think in this context we have an abspath, so I assume this would nearly always be an error if we got here (so I think printing a warning would make more sense, or maybe raising an exception).

scheibelp commented on lib/spack/spack/relocate.py:
```diff
@@ -35,10 +37,77 @@ def _patchelf() -> Optional[executable.Executable]:
+    """Returns the full path to the Windows compiler wrapper if installed"""
+
+    # TODO: Should we query the Spack store instead?
+    # is it worth bootstrapping here?
```
I'm not sure what this comment means: like query the central store instead of the bootstrap store?

I think it's OK (and in fact preferable) to use the bootstrap store (and `ensure_...` functions that call `ensure_executables_in_path_or_raise` will do that when invoked in the context of `ensure_bootstrap_configuration`, which they do in your PR as well as outside of it).

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -819,11 +819,16 @@ class FileTypes:
+        read_size = 8 if not sys.platform == "win32" else 64
+        # first check if this is an ELF, MSVC, or mach-o binary.
+        magic = f.read(read_size)
         if len(magic) < 8:
```
I think this would need to be `if len(magic) < read_size`

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +61,176 @@ def fix_darwin_install_name(path: str) -> None:
+    reloc = relocate(package)
+    lib_map = {}
+    for lib in llnl.util.filesystem.find(package.spec.prefix, "*.lib"):
+        print(f"processing lib {lib}")
```
`tty.msg/debug`?

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +61,176 @@ def fix_darwin_install_name(path: str) -> None:
+    if package:
+        spec = package.spec["compiler-wrapper"]
+    else:
+        spec = spack.store.STORE.db.query_local("compiler-wrapper", installed=True)[0]
```
If we are here, I would prefer to access `bootstrap._msvc_relocate()` (or a public-access version of it).

I don't see any callers passing in `None` for package (e.g. `verify_import_lib` appears to always be called with a not-`None` value).

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +61,176 @@ def fix_darwin_install_name(path: str) -> None:
+    raw_id_data = ctypes.string_at(data_pointer, res_size)
+    str_id_data = raw_id_data.decode(encoding="utf-8").strip("\x00")  # strip null terminator
+    kernel32.FreeLibrary(module_handle)
+    return str_id_data if res_size else None
```
Question: would `None` be a common return value, e.g. if the DLL has no dependencies? Or would it be unexpected for an artifact generated w/ Spack?

If it is unexpected, it might be good to generate debug output here that might give someone a clue that something is going wrong.

johnwparent replied:
This is looking specifically for the spack resource we inject into every spack derived dll, dependencies or no (we use it to map from stage to install prefix). It would be expected for a non spack derived dll to return none here, for example if something were to use the CMake install windows runtime feature and we happened to scan the windows system binaries (which of course would not have the spack resource). 

That said, I think a debug warning couldn't hurt.

scheibelp commented on lib/spack/spack/util/filesystem.py:
```diff
@@ -48,3 +61,176 @@ def fix_darwin_install_name(path: str) -> None:
+    raw_exports = raw_exports[8:]
+    exports = []
+    for export_line in raw_exports:
+        if export_line == "  Summary\r":
```
(minor) Is it always two spaces? Or would regex-matching on `\s+` be more-robust? This seems to have been working quite a while for you, and presumably would only fail if the output of `dumpbin` changed (and I assume there are no options to dumpbin that would influence the stability of its output like `--porcelain`)

johnwparent replied:
It anecdotally is 2 spaces always, that said, there's no actual spec for dumpbin output, so msft is always free to change it. That said if they do, the rest of the string based parsing we have for that output will also probably break, and there's no guarantee that a regex would survive either. 

From a bit of quick research, it seems like the output structure of dumpbin hasn't changed much in the last 20 some odds years, so I feel ok with this, but I'm happy to make it a regex as well.

scheibelp reviewed (changes requested):
I have a few requests and a few questions (all minor I think).

johnwparent committed:
Obtain wrapper from bootstrap

Support case where we're installing from a buildcache but when relocate is unneccesary

the compiler-wrapper is not a link/run dep, so packages can be pulled from a bc without one.
We then need a mechanism to be able to use relocate without having a wrapper already on the system

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
inject msvc environment in relocate runs

Signed-off-by: John Parent <john.parent@kitware.com>

no circular import

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
simplify relocate setup

Signed-off-by: John Parent <john.parent@kitware.com>

Debug wrapper

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Add more debugging

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
wip

Signed-off-by: John Parent <john.parent@kitware.com>

small cleanup

Signed-off-by: John Parent <john.parent@kitware.com>

overhaul relocation pe detection

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
rm debugging statement

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
fixup boundry checking misstep

Signed-off-by: John Parent <john.parent@kitware.com>

rm more debug

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
no more "export" compiler argument

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
debug string

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
try normalizing paths during relocation

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
get more context out of the wrapper

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
remove memoization

Introduced subtle PATH errors

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Use proper lib path for invocation command

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Allow bootstrapped relocate to have correct env

Signed-off-by: John Parent <john.parent@kitware.com>

rm dead code line

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
don't set install prefix during relocate

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Allow non wrapper based packages to skip wrapper relocation

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
restore setting prefix

Signed-off-by: John Parent <john.parent@kitware.com>

scheibelp commented on lib/spack/spack/ci/__init__.py:
```diff
@@ -472,6 +472,7 @@ def generate_pipeline(env: ev.Environment, args) -> None:
         args: (spack.main.SpackArgumentParser): Parsed arguments from the command
             line.
     """
+    cfg.set("concretizer:reuse", False, scope="command_line")
```
Interesting addition. I'm not actually familiar with this strategy of assigning config to the command line. It's probably fine, but generally it's more-idiomatic to use a context manager to create a temporary config scope.

Like

```
    with cfg.override("concretizer:reuse", False):
        with env.write_transaction():
            env.concretize()
            env.write()
```

Also I agree with what this is doing (basically, pipeline generation should be repeatable), but it seems like a significant change not specifically relevant to this PR: it would be ideal to split this off.

johnwparent replied:
Initially I had zero recollection of doing this, but now I remember, we did this when we couldn't figure out the CI's failure to concretize the compiler wrapper when it referred to a git branch version. This can just be wholly dropped.

scheibelp commented on lib/spack/spack/hooks/windows_runtime_linkage.py:
```diff
@@ -4,4 +4,5 @@
 def post_install(spec, explicit=None):
-    spec.package.windows_establish_runtime_linkage()
+    if not spec.package.installed_from_binary_cache:
+        spec.package.windows_establish_runtime_linkage()
```
I can't remember if symlinks are (were) bundled into the binary cache - is that why runtime linkages aren't required?

Once we use the new compiler wrapper, I assume this would also be unnecessary: which I think would be like:

`if not spec.package.installed_from_binary_cache or spec.satisfies(^compiler-wrapper@:x)`, where x is the last version that depends on this

johnwparent replied:
This isn't related to symlinks vs the binary rpath, it's just not needed with the new wrapper (and IIRC, the symlink relocation in the bin cache handles this on the symlink rpath side). 
Binary cache extraction already does the relocation for the binaries, we don't need to essentially redo the operation a second time, which is what this install hook does (its intended for stage->install prefix relocation with the binary rpaths).

scheibelp commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -1939,12 +1958,16 @@ def relocate_package(spec: spack.spec.Spec) -> None:
 
     # Delete identity mappings from prefix_to_prefix
     prefix_to_prefix = {k: v for k, v in prefix_to_prefix.items() if k != v}
+    sfn_prefix_to_prefix = {k: v for k, v in sfn_prefix_to_prefix.items() if k != v}
```
If I understand, the values here are non-SFN paths: I think the filter would have to be updated to properly clean up

johnwparent replied:
We actually don't want to filter here at all since we actually do want "identity mappings" here since it's a sfn to non sfn translation. 
Removing this.

johnwparent committed:
Rm cruft line

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Rm debugging change

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
move relocatinn code to relocate

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
move offending import to local

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Cleanup imports

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
More attempts at import fixing

Signed-off-by: John Parent <john.parent@kitware.com>
